### PR TITLE
feat: WAF-resistant settings save via REST API + JSON (#285)

### LINKED ISSUE #285: [Bug] Settings page returns 403 on hosts with ModSecurity/WAF (OWASP CRS false positives)
## Description

The SlimStat Settings page (`/wp-admin/admin.php?page=slimview-options`) returns **HTTP 403 Forbidden** on hosting environments with ModSecurity/OWASP CRS enabled. The server-level WAF blocks the form POST before it reaches WordPress, preventing users from saving any settings.

This affects shared hosting providers including Namecheap, GoDaddy, Bluehost, SiteGround, and any host running LiteSpeed with ModSecurity.

## Root Cause

The settings form sends all fields via standard `application/x-www-form-urlencoded` POST. Several field values contain patterns that trigger OWASP CRS false positives:

| Field | Content | CRS Rules Triggered |
|-------|---------|-------------------|
| `options[custom_css]` | CSS code (`{ }`, `/*`, `*/`) | 941xxx (XSS), 942xxx (SQLi) |
| `options[opt_out_message]` | HTML markup (`<a href>`, `<p>`) | 941100-941320 (XSS) |
| `options[opt_out_cookie_names]` | JSON-like `key=value` | 942xxx (SQLi) |
| `options[ignore_referers]` | Full URLs (`https://...`) | 934xxx (SSRF), 931130 (RFI) |
| `options[external_domains]` | Full URLs | 934xxx (SSRF) |
| `options[rest_api_tokens]` | Random special chars | 942xxx (SQLi) |

A single CRITICAL rule match (5 points) exceeds the PL1 blocking threshold, triggering rule **949110** (Inbound Anomaly Score Exceeded).

The CRS WordPress exclusion rules (`REQUEST-903.9002`) only cover vanilla WordPress — third-party plugins are not included and this file is disabled by default on most shared hosts.

## Steps to Reproduce

1. Install SlimStat on a host with ModSecurity/OWASP CRS (e.g., Namecheap shared hosting)
2. Navigate to SlimStat → Settings
3. Click "Save" on any tab containing HTML/CSS/URL fields (especially Tracker or Reports tab)
4. **Expected**: Settings saved successfully
5. **Actual**: HTTP 403 Forbidden (server-level block page, not WordPress error)

## Impact

- Users cannot configure SlimStat at all on affected hosts
- Default settings may not match user's needs (GDPR consent, tracking method, exclusions)
- Support burden: recurring tickets from shared hosting users
- First reported: Ticket #15057 (Namecheap, LiteSpeed + ModSecurity)

## Environment

- SlimStat: 5.4.8+ (all versions affected — form POST pattern unchanged)
- WordPress: Any
- PHP: Any
- Server: LiteSpeed / Apache with ModSecurity + OWASP CRS
- Hosting: Namecheap (confirmed), likely GoDaddy, Bluehost, SiteGround, InMotion

## Proposed Solution

Migrate settings save from form POST to REST API with JSON body, with base64 fallback and WAF auto-detection. Full 4-phase implementation plan posted as the first comment below.

## Research

Full analysis with 70+ sources available in the research document covering:
- Field-by-field WAF trigger analysis
- OWASP CRS rule mapping (941xxx XSS, 942xxx SQLi, 934xxx SSRF)
- Proven patterns from Beaver Builder (base64), AIOSEO (REST+JSON), Gutenberg (JSON Content-Type)
- Hosting landscape (Namecheap, GoDaddy, SiteGround, Bluehost ModSecurity configs)
- No other WordPress analytics plugin has built-in WAF mitigation

### PR DESCRIPTION
**Pull request by parhumm**
## Summary

- Extract settings save logic into reusable `SettingsSaveService` for both form POST and REST API paths
- Add `POST /slimstat/v1/settings` REST endpoint that accepts JSON body (avoids ModSecurity/OWASP CRS false positives)
- JS form intercept with automatic base64 fallback if WAF blocks JSON
- WAF auto-detection service with admin notice on settings page

## Problem

The Settings page returns **HTTP 403 Forbidden** on hosts with ModSecurity/OWASP CRS (Namecheap, GoDaddy, Bluehost, SiteGround, LiteSpeed). The form POST contains HTML (`opt_out_message`), CSS (`custom_css`), URLs, and JSON-like patterns that trigger XSS (941xxx) and SQLi (942xxx) CRS rules.

No other WordPress analytics plugin has built-in WAF mitigation.

## Solution

4-phase approach based on proven patterns from Beaver Builder (base64) and AIOSEO (REST API + JSON):

1. **Phase 1**: `SettingsSaveService` — shared save logic (pure refactor, no behavior change)
2. **Phase 2**: `SettingsRestController` — REST endpoint + probe endpoint
3. **Phase 3**: JS form intercept → `fetch()` with JSON body, base64 retry on 403
4. **Phase 4**: `WafDetectionService` — loopback probe with 24h cache + admin notice

## Graceful Degradation

- JS disabled → traditional form POST still works
- REST API blocked → falls back to form POST
- JSON blocked by WAF → retries with base64 encoding
- Base64 blocked → falls back to traditional form POST

## Test plan

- [ ] PHPUnit: 66 tests, 102 assertions (identical to baseline)
- [ ] Custom tests: all pass (rest-controller, browscap-isolation, v5460-migration, gdpr-consent)
- [ ] PHP syntax check: clean on all 7 files
- [ ] Manual: Save settings on each tab via browser
- [ ] See test checklist and QA checklist comments below

Fixes #285

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * REST-based settings save with server-probe to detect blocking, retry fallback, and improved in-page save UX and dismissible admin notices.
* **Bug Fixes**
  * More robust, unified settings persistence (including multisite/network handling) and clearer warning messages when saves fail or are blocked.
* **Tests**
  * New test suites validating REST endpoints and centralized settings-save behavior across various payloads and edge cases.

### CODE REVIEW

**Comment by parhumm** [comment]
## Test Checklist (Updated after review fixes)

### Automated Tests (all passing)

- [x] PHPUnit: 66 tests, 102 assertions — identical to pre-fix baseline
- [x] test:rest-controller — 16 assertions passed
- [x] test:browscap-isolation — 14 assertions passed
- [x] test:v5460-migration — 52 assertions passed
- [x] test:gdpr-consent — 26 assertions passed
- [x] test:settings-save-service — 33 assertions (includes network-admin scenarios)
- [x] test:settings-rest-controller — 20 assertions
- [x] test:waf-settings-bugfix — 14 assertions (Issues 1-5 regression tests)
- [x] PHP syntax check clean on all modified files
- [x] CI: PHP 7.4 + PHP 8.2 tiers pass

Manual test matrix moved to a consolidated checklist: https://github.com/wp-slimstat/wp-slimstat/pull/286#issuecomment-4191796180

**Comment by parhumm** [comment]
## Risks, Side Effects (Updated)

### Known Risks & Side Effects

| Risk | Impact | Mitigation | Status |
|------|--------|-----------|--------|
| Page reloads after AJAX save | Slightly slower than pure AJAX but shows correct form values | WAF benefit preserved (save goes through REST); reload shows PHP notices with proper HTML | Fixed |
| REST API disabled by security plugin | JS intercept fails on fetch() | Non-OK responses (500, 401, etc.) fall through to .catch() then legacy form POST | Fixed |
| WAF probe adds latency | One HTTP loopback request (~100ms) on first settings page load | Cached 24h via transient; only sends auth cookies (not full jar); skips cross-origin | Fixed |
| slimstat_save_options filter context | Pro hooks may expect get_current_screen() which is NULL in REST | wp_slimstat::$save_context provides tab, is_network, and via metadata | Fixed |
| is_network_admin() in REST context | Returns false, breaking multisite | Explicit is_network param passed from JS to REST to service; update_site_option() used correctly | Fixed |
| Pro addon field sanitization | REST path starts with empty base definitions | SettingsSaveService has built-in constants for free fields; Pro fields come via slimstat_options_on_page filter | Mitigated |
| Base64 not guaranteed | CRS v4 auto-decoding can strip base64 | 3-tier fallback: JSON REST then base64 REST then legacy form POST | By design |
| WAF retry: 406/503 | Previously only retried on 403 | Now retries on 403, 406, and 503 | Fixed |

Manual QA checklist moved to a consolidated comment: https://github.com/wp-slimstat/wp-slimstat/pull/286#issuecomment-4191796180

File: admin/assets/js/admin.js [hunk 1]
```diff
@@ -6,6 +6,170 @@ if (typeof SlimStatAdminParams == "undefined") {
...
+        })
+            .then(function (response) {
+                if (response.status === 403) {
+                    // WAF blocked the JSON request — retry with base64
+                    if (!useBase64) {
+                        window.slimstatWafBlocked = true;
+                        return slimstatSaveViaRest(tab, options);
+                    }
+                    throw new Error("WAF blocked");
+                }
```
[the remark this answers is not in the document]
Comment (parhumm) [reply admin/assets/js/admin.js:141]: Fixed in c485f09b. Now checks `response.ok` first — 403/406/503 trigger base64 retry, all other non-OK responses throw to `.catch()` → legacy form POST fallback.

File: src/Controllers/Rest/SettingsRestController.php [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+ *
+ * Provides a WAF-resistant settings save endpoint using JSON body instead
+ * of form-encoded POST, avoiding ModSecurity/OWASP CRS false positives.
+ *
+ * @package   SlimStat\Controllers\Rest
+ * @since     5.4.10
+ * @see       https://github.com/wp-slimstat/wp-slimstat/issues/285
+ */
+
+declare(strict_types=1);
```
[the remark this answers is not in the document]
Comment (parhumm) [reply src/Controllers/Rest/SettingsRestController.php:22]: Fixed in c485f09b. GPL header added and all error strings wrapped with `__('...', 'wp-slimstat')`.

File: src/Controllers/Rest/SettingsRestController.php [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+            }
+        } else {
+            $options = $request->get_param('options');
+            if (!is_array($options)) {
+                return new \WP_REST_Response(['success' => false, 'error' => 'Options must be an object.'], 400);
+            }
+        }
+
+        // Save via the shared service (uses built-in field type knowledge for REST calls)
+        $result = SettingsSaveService::save($tab, $options);
```
[the remark this answers is not in the document]
Comment (parhumm) [reply src/Controllers/Rest/SettingsRestController.php]: Fixed in 7015e61a. REST controller now calls `apply_filters('slimstat_options_on_page', [])` before save, so Pro addon fields get proper type-aware sanitization. Base free plugin definitions use built-in constants (RICH_TEXT_FIELDS, CODE_EDITOR_FIELDS) as fallback. Documented in code comment for future extraction into shared builder.

File: src/Services/SettingsSaveService.php [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+            // Network admin override flags
+            if (is_network_admin()) {
+                if ('on' == ($options['addon_network_settings_' . $slug] ?? 'no')) {
+                    \wp_slimstat::$settings['addon_network_settings_' . $slug] = 'on';
+                } else {
+                    \wp_slimstat::$settings['addon_network_settings_' . $slug] = 'no';
+                }
+            } elseif (isset(\wp_slimstat::$settings['addon_network_settings_' . $slug])) {
+                unset(\wp_slimstat::$settings['addon_network_settings_' . $slug]);
+            }
```
[the remark this answers is not in the document]
Comment (parhumm) [reply src/Services/SettingsSaveService.php:203]: Fixed in 7015e61a. Added `strpos($slug, 'addon_network_settings_') === 0` guard at line 160 of SettingsSaveService.php — these prefixed keys are now skipped from the generic save loop and only processed in the network admin block.

File: src/Services/WafDetectionService.php [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        $response = wp_remote_post($probe_url, [
+            'body'      => wp_json_encode(['test' => '<script>alert(1)</script> SELECT * FROM']),
+            'headers'   => [
+                'Content-Type' => 'application/json',
+                'X-WP-Nonce'   => wp_create_nonce('wp_rest'),
+            ],
+            'timeout'   => 10,
+            'sslverify' => apply_filters('https_local_ssl_verify', false),
+            'cookies'   => wp_unslash($_COOKIE),
+        ]);
```
[the remark this answers is not in the document]
Comment (parhumm) [reply src/Services/WafDetectionService.php:97]: Fixed in c485f09b. Now compares full origin (scheme+host+port) via `wp_parse_url`, and only forwards `AUTH_COOKIE`, `SECURE_AUTH_COOKIE`, `LOGGED_IN_COOKIE` — not the entire cookie jar. Cross-origin probes send no cookies.

File: tests/settings-rest-controller-test.php [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+if (!function_exists('register_rest_route')) {
+    function register_rest_route($namespace, $route, $args) {
+        // Capture registered routes for testing
+        global $registered_routes;
+        $registered_routes[] = ['namespace' => $namespace, 'route' => $route, 'args' => $args];
+    }
+}
+if (!function_exists('current_user_can')) {
+    function current_user_can($cap) { return true; }
+}
```
[the remark this answers is not in the document]
Comment (parhumm) [reply tests/settings-rest-controller-test.php:83]: Acknowledged but not converting in this PR. All 22+ project test files use the same standalone stub pattern — there is no WP_UnitTestCase bootstrap infrastructure. Converting a single test requires setting up the full WP test harness (wp-tests-config.php, test DB, phpunit bootstrap). This is a separate infrastructure initiative tracked for a future PR.

File: src/Services/SettingsSaveService.php [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+ * @since     5.4.10
+ */
+
+declare(strict_types=1);
+
+namespace SlimStat\Services;
+
+if (!defined('ABSPATH')) {
+    exit;
+}
```
[the remark this answers is not in the document]
Comment (parhumm) [reply src/Services/SettingsSaveService.php:27]: Fixed in 49bf17d2. GPL header added to SettingsSaveService.php.

File: wp-slimstat.php [hunk 8]
```diff
@@ -1245,12 +1255,12 @@ public static function init_options()
...
-    public static function update_option($_key = '', $_value = '')
+    public static function update_option($_key = '', $_value = '', bool $_is_network = false)
     {
-        if (!is_network_admin()) {
-            update_option($_key, $_value);
-        } else {
+        if ($_is_network || is_network_admin()) {
             update_site_option($_key, $_value);
+        } else {
+            update_option($_key, $_value);
```
[the remark this answers is not in the document]
Comment (parhumm) [reply wp-slimstat.php:1268]: Fixed in 49bf17d2. `update_option()` now enforces `current_user_can('manage_network_options')` before calling `update_site_option()`. Non-super-admins attempting network-scoped writes are silently rejected.

File: admin/assets/js/admin.js [hunk 9]
```diff
@@ -6,6 +6,169 @@ if (typeof SlimStatAdminParams == "undefined") {
...
+            .then(function (response) {
+                // 403, 406, 503 are common WAF-block status codes
+                if (response.status === 403 || response.status === 406 || response.status === 503) {
+                    if (!useBase64) {
+                        window.slimstatWafBlocked = true;
+                        return slimstatSaveViaRest(tab, options);
+                    }
+                    throw new Error("WAF blocked");
+                }
+                return response.json();
```
[the remark this answers is not in the document]
Comment (parhumm) [reply admin/assets/js/admin.js:142]: Fixed in c485f09b. Now checks `response.ok` first. 403/406/503 trigger WAF base64 retry; all other non-OK responses (401, 404, 500, etc.) throw to `.catch()` which calls `form.submit()` for legacy fallback.

File: src/Services/WafDetectionService.php [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+/**
+ * WAF Detection Service
+ *
+ * Probes the REST API settings endpoint to detect if a server-level WAF
+ * (ModSecurity, LiteSpeed, Cloudflare, etc.) is blocking requests.
+ *
+ * @package   SlimStat\Services
+ * @since     5.4.10
+ * @see       https://github.com/wp-slimstat/wp-slimstat/issues/285
+ */
```
[the remark this answers is not in the document]
Comment (parhumm) [reply src/Services/WafDetectionService.php:20]: Fixed in c485f09b. GPL header added to WafDetectionService.php.

**Comment by parhumm** [comment]
## Human QA Checklist (Consolidated)

> Combines manual test items from the risks/side-effects analysis and the test checklist into a single deduplicated list. Run all CRITICAL items before merge; HIGH and MEDIUM items can be split across reviewers.

### Priority: CRITICAL (must verify before merge)

**Save flow & sanitization**

- [ ] **1. All 6 free tabs save end-to-end** — General, Tracker, Reports, Exclusions, Access Control, Maintenance. Verify every field type (toggle, text, integer, select, textarea, rich_text) persists correctly on its respective tab.
- [ ] **2. Toggle ON → OFF persistence** — Turn `is_tracking` ON, save, then OFF, save. Verify `"no"` is correctly sent and persisted (toggle hidden field pattern works through REST).
- [ ] **3. Custom CSS with WAF-trigger patterns** (Tab 3) — Enter `.report { color: red; } /* comment */ url("test")`. Verify persisted exactly via `wp_strip_all_tags()`.
- [ ] **4. HTML in `opt_out_message`** (Tab 2) — Enter `<p>This site uses <a href="/privacy">cookies</a>.</p>`. Verify HTML preserved via `wp_kses_post()`.
- [ ] **5. Cookie names with `key=value` patterns** (Tab 2) — Save `opt_out_cookie_names` and `opt_in_cookie_names` with values like `CookiePreferences-site.com="slimstat":false`. Verify exact persistence.
- [ ] **6. URLs in `ignore_referers`** (Tab 4) — Save `https://mysite.com*` patterns. Verify full URLs persist.
- [ ] **7. URLs in `external_domains`** (Tab 2) — Save full HTTPS URLs. Verify persistence.

**Fallback paths**

- [ ] **8. JavaScript disabled** — Disable JS in browser → save on Tab 1 → verify settings persist via traditional form POST (no REST involved).
- [ ] **9. REST API blocked / non-WAF error** — Either install a security plugin that blocks REST, or temporarily return 500 from the endpoint. Verify `.catch()` fires and form submits traditionally (not stuck).
- [ ] **10. Page reload after save** — After successful save, verify page reloads and shows the green PHP-rendered admin notice (not a JS-injected one).

### Priority: HIGH

**WAF detection & retry**

- [ ] **11. Simulate 403 response** — Force REST endpoint to return 403 → verify auto-retry with base64 encoding kicks in.
- [ ] **12. Simulate 406 response** — Verify base64 retry triggers (regression for fixed comment).
- [ ] **13. Simulate 503 response** — Verify base64 retry triggers.
- [ ] **14. WAF admin notice** — Block the probe endpoint with a firewall rule. Verify yellow warning notice appears identifying the WAF type (LiteSpeed/ModSecurity/Cloudflare).
- [ ] **15. WAF probe on clean host** — On a host without a WAF, verify no admin notice is shown and the probe doesn't fail.
- [ ] **16. WAF probe behind Cloudflare** — Test on a Cloudflare-proxied site. Verify probe doesn't hang and doesn't leak cookies cross-origin (cookie jar should only include AUTH/SECURE_AUTH/LOGGED_IN cookies if same-origin).

**Special-case handlers**

- [ ] **17. DB indexes ON** (Tab 6) — Toggle ON → verify 4 indexes created on `slim_stats` table (`stats_resource_idx`, `stats_browser_idx`, `stats_searchterms_idx`, `stats_fingerprint_idx`). Success message contains HTML link.
- [ ] **18. DB indexes OFF** (Tab 6) — Toggle OFF → verify all 4 indexes dropped. Message confirms removal.
- [ ] **19. Geolocation provider change** — Change from `dbip` to `maxmind` → save → verify `wp_slimstat_update_geoip_database` cron is scheduled (check `wp_options` for `cron`).
- [ ] **20. Browscap enable** — Toggle browscap ON → verify browscap database download is triggered and admin notice reflects success/failure.
- [ ] **21. Tracking method change** — Switch between `rest`/`ajax`/`adblock_bypass` → save → verify `slimstat_permalink_structure_updated` transient set and rewrite rules flushed on next page load.

**Multisite**

- [ ] **22. Network admin save** — On multisite, save from Network Admin → verify `addon_network_settings_*` flags set and `update_site_option()` is used (check `wp_sitemeta` table, NOT `wp_options`).
- [ ] **23. JS sends `is_network: true`** — On multisite Network Admin, open DevTools Network tab → verify the REST request body includes `"is_network": true`.
- [ ] **24. Capability enforcement** — As a non-super-admin user, attempt to call the REST endpoint with `is_network: true` → verify the request is silently rejected (no `update_site_option()` call) due to `manage_network_options` check.

**DevTools verification**

- [ ] **25. REST request inspection** — Open DevTools → Network → save settings on any tab → verify `fetch()` to `/wp-json/slimstat/v1/settings` with `Content-Type: application/json` and JSON body containing `{tab, options, nonce, is_network}`.

### Priority: MEDIUM

**Pro addon compatibility** (only if Pro is installed)

- [ ] **26. Pro addon settings on tab 7/8** — Save Pro settings (heatmap, custom DB, export, user overview). Verify they persist via the same REST path.
- [ ] **27. Pro heatmap custom CSS** — Enter CSS in Pro's heatmap custom CSS field → save tab 7. Verify CSS preserved (not mangled by `sanitize_text_field`).
- [ ] **28. Pro license validation** — On tab 8, save license key. Verify Pro's `slimstat_save_options` hook runs and `wp_slimstat::$save_context` provides correct tab/is_network/via metadata to the validator.

**Edge cases**

- [ ] **29. Large custom CSS (>10KB)** — Paste 10KB+ of CSS → save → verify full content persisted (no truncation by REST body limits).
- [ ] **30. `SLIMSTAT_MODSEC_FIX` constant** — Add `define('SLIMSTAT_MODSEC_FIX', true);` to `wp-config.php` → verify base64 is used from the first save (no retry round-trip needed).
- [ ] **31. Two admins save simultaneously** — Open settings in two browser tabs → change different settings in each → save both. Verify no crash; last write wins.

File: src/Controllers/Rest/SettingsRestController.php [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+     * Register REST routes for settings save and WAF probe.
+     */
+    public function register_routes(): void
+    {
+        register_rest_route('slimstat/v1', '/settings', [
+            'methods'             => 'POST',
+            'callback'            => [$this, 'save_settings'],
+            'permission_callback' => static function () {
+                return current_user_can('manage_options');
+            },
```
[the remark this answers is not in the document]
Comment (parhumm) [reply src/Controllers/Rest/SettingsRestController.php:45]: Addressed in a1011c02. Added explicit upfront capability check in save_settings() that returns 403 when is_network=true and the user lacks manage_network_options.
